### ISSUE TITLE
Add option to control whether multiline properties in Objects are pretty printed with blank lines

### ISSUE THREAD

**Issue by asselin** [benjamn/recast#228]
Currently, multiline properties in Objects are pretty printed with blank lines before/after them.  For example:
````
        $stateProvider
            .state('map', {
                parent: 'root',
                url: '/map',

                views: {
                    'content@app': {
                        templateUrl: 'app/map/map.html',
                        controller: 'MapController',
                        controllerAs: 'vm'
                    }
                }
            });
````

Add an option to control whether the extra blank line is inserted.  For example, setting to false would produce:
````
        $stateProvider
            .state('map', {
                parent: 'root',
                url: '/map',
                views: {
                    'content@app': {
                        templateUrl: 'app/map/map.html',
                        controller: 'MapController',
                        controllerAs: 'vm'
                    }
                }
            });
````

**Comment by benjamn** [comment]
There are a couple of distinct considerations here:
1. how the pretty-printer should print entirely new code (related: #226)
2. how object properties should be "reprinted" when only some of the properties have changed

Re: 2., I think there's an opportunity here to do what [`printStatementSequence`](https://github.com/benjamn/recast/blob/060b6fc9df7956a6f39df4444a04740c8e57621d/lib/printer.js#L1554-L1602) does for statements, so whitespace between pre-existing object properties is preserved, rather than blown away. Then you wouldn't need to rely on a dedicated printer option to achieve the goal of preserving existing whitespace.

Re: 1., I like to avoid accumulating special-purpose options if possible, so I think it's worth asking if we should just change the printer so that it always puts newlines around multi-line properties (when printing entirely new code). That's [how it works for statements](https://github.com/benjamn/recast/blob/060b6fc9df7956a6f39df4444a04740c8e57621d/lib/printer.js#L1573) currently. Unless you can think of situations where you would really want there to be no newlines around properties (throughout the entire file), then let's just change the printer behavior to prefer spaces and not introduce a new option?

**Comment by asselin** [comment]
@benjamn I agree on your 2nd point.  For the first point, this is a matter of preference and possibly coding standards, so I firmly believe it needs to be an option.  However, I do agree with your point about not making a ton of special purpose options.

I looked at .eslintrc, and didn't see a rule related to this-- if there was, I'd say this could be the first option implemented in the formatter.  I suppose it could still be part of that infrastructure, and it could be a recast extension.  Any suggestions on how you'd like to see options like this implemented?

**Comment by qimingweng** [comment]
@asselin @benjamn Also agreed on point number 2, but regarding number 1:

Are new lines around multiline object key declarations standard in popular js style guides? They seem a little bit unnecessary to me and unexpected from defaults. I took a look at the eslint documentation and couldn't find any rules regarding this in general (or at least no rules on this page http://eslint.org/docs/rules/). I would almost say that it makes more sense not to add new lines as the only option or the default.

**Comment by asselin** [comment]
@qimingweng I tend to agree, but I wanted to submit a PR that didn't change the current default behavior.

**Comment by xmo-odoo** [comment]
Has there been any progress on part 2? It's not *hard* but it's a bit frustrating to rollback newlines added between existing (unchanged) object properties.
